File: patterns/java/src/refactoring_guru/patterns/iterator/example/social_networks/LinkedIn.java
package refactoring_guru.patterns.iterator.example.social_networks;

import refactoring_guru.patterns.iterator.example.iterators.LinkedInIterator;
import refactoring_guru.patterns.iterator.example.iterators.ProfileIterator;
import refactoring_guru.patterns.iterator.example.profile.Profile;

import java.util.ArrayList;
import java.util.List;

public class LinkedIn implements SocialNetwork {
    private List<Profile> contacts;

    public LinkedIn(List<Profile> cache) {
        if (cache != null) {
            this.contacts = cache;
        } else {
            this.contacts = new ArrayList<>();
        }
    }

    public Profile requestContactInfoFromLinkedInAPI(String profileEmail) {
        // Here would be a POST request to one of the LinkedIn API endpoints.
        // Instead, we emulates long network connection, which you would expect in the real life...
        simulateNetworkLatency();
        System.out.println("LinkedIn: Loading profile '" + profileEmail + "' over the network...");

        // ...and return test data.
        return findContact(profileEmail);
    }

    public List<String> requestRelatedContactsFromLinkedInAPI(String profileEmail, String contactType) {
        // Here would be a POST request to one of the LinkedIn API endpoints.
        // Instead, we emulates long network connection, which you would expect in the real life.
        simulateNetworkLatency();
        System.out.println("LinkedIn: Loading '" + contactType + "' list of '" + profileEmail + "' over the network...");

        // ...and return test data.
        Profile profile = findContact(profileEmail);
        if (profile != null) {
            return profile.getContacts(contactType);
        }
        return null;
    }

    private Profile findContact(String profileEmail) {
        for (Profile profile : contacts) {
            if (profile.getEmail().equals(profileEmail)) {
                return profile;
            }
        }
        return null;
    }

    private void simulateNetworkLatency() {
        try {
            Thread.currentThread().sleep(2500);
        } catch (InterruptedException ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public ProfileIterator getFriendsIterator(String profileEmail) {
        return new LinkedInIterator(this, "friends", profileEmail);
    }

    @Override
    public ProfileIterator getCoworkersIterator(String profileEmail) {
        return new LinkedInIterator(this, "coworkers", profileEmail);
    }
}

File: patterns/cs/src/RefactoringGuru/Iterator/Example/SocialNetworks/LinkedIn.cs
using System;
using System.Collections.Generic;
using System.Threading;
using RefactoringGuru.Iterator.Example.Iterators;
using RefactoringGuru.Iterator.Example.Profiles;

namespace RefactoringGuru.Iterator.Example.SocialNetworks
{
    class LinkedIn : SocialNetwork
    {
        private List<Profile> contacts;

        public LinkedIn(List<Profile> cache)
        {
            if (cache != null)
            {
                this.contacts = cache;
            }
            else
            {
                this.contacts = new List<Profile>();
            }
        }

        public Profile RequestContactInfoFromLinkedInAPI(String profileEmail)
        {
            // EN: Here would be a POST request to one of the LinkedIn API
            // endpoints. Instead, we emulates long network connection, which
            // you would expect in the real life...
            // 
            // RU: Здесь бы был POST запрос к одному из адресов API LinkedIn. Но
            // вместо этого мы эмулируем долгое сетевое соединение, прямо как в
            // реальной жизни...
            SimulateNetworkLatency();
            Console.WriteLine("LinkedIn: Loading profile '" + profileEmail + "' over the network...");

            // EN: ...and return test data.
            // 
            // RU: ...и возвращаем тестовые данные.
            return FindContact(profileEmail);
        }

        public List<String> RequestRelatedContactsFromLinkedInAPI(String profileEmail, String contactType)
        {
            // EN: Here would be a POST request to one of the LinkedIn API
            // endpoints. Instead, we emulates long network connection, which
            // you would expect in the real life.
            // 
            // RU: Здесь бы был POST запрос к одному из адресов API LinkedIn. Но
            // вместо этого мы эмулируем долгое сетевое соединение, прямо как в
            // реальной жизни...
            SimulateNetworkLatency();
            Console.WriteLine("LinkedIn: Loading '" + contactType + "' list of '" + profileEmail + "' over the network...");

            // EN: ...and return test data.
            // 
            // RU: ...и возвращаем тестовые данные.
            Profile profile = FindContact(profileEmail);
            if (profile != null)
            {
                return profile.GetContacts(contactType);
            }
            return null;
        }

        private Profile FindContact(String profileEmail)
        {
            foreach (var profile in contacts)
            {
                if (profile.GetEmail().Equals(profileEmail))
                {
                    return profile;
                }
            }
            return null;
        }

        private void SimulateNetworkLatency()
        {
            Thread.Sleep(2500);
        }

        public ProfileIterator GetFriendsIterator(string profileEmail)
        {
            return new LinkedInIterator(this, "friends", profileEmail);
        }

        public ProfileIterator GetCoworkersIterator(string profileEmail)
        {
            return new LinkedInIterator(this, "coworkers", profileEmail);
        }
    }
}
